src/ucs/datastruct/ptr_array.h

```diff
@@ -121,12 +135,133 @@ void *ucs_ptr_array_replace(ucs_ptr_array_t *ptr_array, unsigned index, void *ne
+
+
+/*
+   Locked interface
+*/
+
+/* Locked ptr array */
+typedef struct ucs_ptr_array_locked {
+    ucs_ptr_array_t super;
+    ucs_recursive_spinlock_t lock;
+} ucs_ptr_array_locked_t;
+
+/**
+ * Locked array init
+ */
+void ucs_ptr_array_locked_init(ucs_ptr_array_locked_t *locked_ptr_array, uint32_t init_placeholder,
+                        const char *name);
+
+
+/**
+ * Cleanup the locked array.
+ *
+ * @param locked_ptr_array    Pointer a locked ptr array.
+ *
+ * All values should already be removed from it.
+ */
+void ucs_ptr_array_locked_cleanup(ucs_ptr_array_locked_t *locked_ptr_array);
+
+
+/**
+ * Insert a pointer to the locked array.
+ *
+ * @param value        Pointer to insert. Must be 8-byte aligned.
+ * @param placeholder  Filled with placeholder value.
+ * @return             The index to which the value was inserted.
+ *
+ * Complexity: amortized O(1)
+ *
+ * Note: The array will grow if needed.
+ */
+unsigned ucs_ptr_array_locked_insert(ucs_ptr_array_locked_t *locked_ptr_array, void *value,         
+                   uint32_t *placeholder_p);
 
 
+/**
+ * Remove a pointer from the locked array.
+ *
+ * @param locked_ptr_array  Pointer a locked ptr array.
+ * @param index             Index to remove from.
+ * @param placeholder       Value to put in the free slot.
+ *
+ * Complexity: O(1)
+ */
+void ucs_ptr_array_locked_remove(ucs_ptr_array_locked_t *locked_ptr_array, unsigned index,
+                          uint32_t placeholder);
+
+
+/**
+ * Replace pointer in the locked array
+ *
+ * @param  index    index of slot
+ * @param  new_val  value to put into slot given by index
+ * @return old value of the slot
+ *
+ * Complexity: O(1)
+ */
+void *ucs_ptr_array_locked_replace(ucs_ptr_array_locked_t *locked_ptr_array,
+	unsigned index, void *new_val);
+
+
+/* Acquire ptr_array lock */
+#define ucs_ptr_array_locked_acquire_lock(_locked_ptr_array)\
+    ucs_recursive_spin_lock(&(_locked_ptr_array)->lock)
+
+
+/* Release ptr_array lock */
+#define ucs_ptr_array_locked_release_lock(_locked_ptr_array)\
+    ucs_recursive_spin_unlock(&(_locked_ptr_array)->lock)
+
+
+/**
+ * Retrieve a value from the locked array.
+ *
+ * The user is required to lock using ucs_ptr_array_locked_acquire_lock()
+ * and ucs_ptr_array_locked_release_lock() functions.
+ *
+ * @param index   Index to retrieve the value from.
+ * @param value   Filled with the value.
+ * @return        Whether the value is present and valid.
+ *
+ * Complexity: O(1)
+ */
+#define ucs_ptr_array_locked_lookup(_locked_ptr_array, _index, _var)\
+    ucs_ptr_array_lookup(&(_locked_ptr_array)->super, _index, _var)
+
+/**
+ * Iterate over all valid elements in the locked array.
+ * The user is required to lock using ucs_ptr_array_locked_acquire_lock()
+ * and ucs_ptr_array_locked_release_lock() functions.
+ *
+ * @param _var    Pointer to current array element in the foreach.
```

Comment:
pls add
`[in]`/`[out]`/`[in/out]` for paramateres as requested above

Reply:
Fixed all.
Hi
@dmitrygx , Regarding some of the above 'not fixed yet' comments (if they are still not fixed in the commit I will do in a few minutes) - Could you please explain what needs to be fixed?